### PR TITLE
[stable/traefik] Allow DNS challenge to be used for getting Letsencrypt Certificate

### PR DESCRIPTION
**Pull request by kevinjqiu**

### CODE REVIEW

**Comment by viglesiasce** [comment]
@k8s-bot ok to test

**Comment by viglesiasce** [comment]
This fails to install with:

```
 $  helm install stable/traefik/
Error: YAML parse error on traefik/templates/deployment.yaml: error converting YAML to JSON: yaml: line 48: mapping values are not allowed in this context
```

**Comment by viglesiasce** [comment]
@kevinjqiu thanks for resolving the merge conflicts!!

**Comment by krancour** [comment]
This is a feature I've wanted to see added for some time! Reviewing...

File: stable/traefik/templates/deployment.yaml
Comment (krancour) [inline stable/traefik/templates/deployment.yaml]: This has me a little concerned. I know this is the easy way to shim a whole bunch of environment variables (which will vary from one DNS provider to the next) into the pods, _but_... we're mainly talking about username/passwords or secret keys, etc. here. i.e. All things that really belong in a secret. So, while this works, I think there's some less-than-ideal exposure of credentials here.

Comment (kevinjqiu) [reply stable/traefik/templates/deployment.yaml]: @krancour Good point. I will give it another thought.

Comment (kevinjqiu) [reply stable/traefik/templates/deployment.yaml]: @krancour ok I refreshed my memory on this (it's been a while :)). My original intent was to have the sensitive environment variables be drawn from secrets. e.g.:

```
  - name: AWS_REGION
    value: us-east-1
  - name: AWS_ACCESS_KEY_ID
    valueFrom:
        secretKeyRef:
            name: aws
            key: ACCESS_KEY_ID
  - name: AWS_SECRET_ACCESS_KEY
    valueFrom:
        secretKeyRef:
            name: aws
            key: SECRET_ACCESS_KEY
```

Maybe I can document the intended usage in `templates/NOTES.txt`?

**Comment by unguiculus** [comment]
ping @krancour

**Comment by krancour** [comment]
Sorry... I lost track of this one. I promise to look at this more closely before the end of the week.

**Comment by gdmello** [comment]
@krancour - This is a useful feature - any chance of a review?

**Comment by unguiculus** [comment]
Marking this as stale. Please update within one week.

**Comment by kevinjqiu** [comment]
@unguiculus Is it going to be looked at if I resolve conflict with master? I mean no disrespect but [multiple](https://github.com/kubernetes/charts/pull/1299#discussion_r126038450) [previous](https://github.com/kubernetes/charts/pull/1299#issuecomment-321061449) [calls](https://github.com/kubernetes/charts/pull/1299#issuecomment-328905942) for review and discussion have been responded with unresponsiveness...I recognize fully the preciousness of time of the maintainers but I also don't want to waste my own time...

Anyway, happy to update as long as there's an owner for this PR.

**Comment by krancour** [comment]
@kevinjqiu, I take a lot of the blame for that. You hit the nail on the head with acknowledging everyone is busy-- and your own time is no less precious.

imo, let the decision to invest more time in this or not hinge on whether this feature is (still) valuable to yourself. If you decide it is, I will commit to reviewing. You can even DM me on Twitter (same handle as here) to nag me if I don't follow through promptly.

**Comment by kevinjqiu** [comment]
> and your own time is no less precious

@krancour thanks for acknowledging that.

> on whether this feature is (still) valuable to yourself

Yes, we have been using this in prod for a couple of months. We setup our
own helm repository for a forked version of this chart including this
change together with a few other features.

As stated, I'm more than happy to bring this branch up-to-date with the
latest master as long as it's going to be looked at.





On Nov 20, 2017 5:40 AM, "Kent Rancourt" <notifications@github.com> wrote:

@kevinjqiu <https://github.com/kevinjqiu>, I take a lot of the blame for
that. You hit the nail on the head with acknowledging everyone is busy--
and your own time is no less precious.

imo, let the decision to invest more time in this or not hinge on whether
this feature is (still) valuable to yourself. If you decide it is, I will
commit to reviewing. You can even DM me on Twitter (same handle as here) to
nag me if I don't follow through promptly.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/charts/pull/1299#issuecomment-345698197>,
or mute
the thread

File: stable/traefik/templates/deployment.yaml
Comment (krancour) [reply stable/traefik/templates/deployment.yaml]: And those secrets would be managed separately from the chart? tbh, that doesn't seem like an ideal user experience.

You're enabling a great degree of flexibility here by allowing freeform entry of env variables and values (or values from secrets), but it seems like lots of rope to hang oneself with, no?

I know from a contribution perspective, it probably seems less appealing to build in specific support for each DNS provider, but imho, that's the way to go if you want a maximally optimal UX for users of this feature.

If the above is too daunting a proposition, maybe it's worth it to add support for a few of the most common DNS providers first and others can be contributed over time?

**Review by krancour — CHANGES_REQUESTED** [review]

File: stable/traefik/templates/deployment.yaml
Comment (kevinjqiu) [reply stable/traefik/templates/deployment.yaml]: @krancour 

> I know from a contribution perspective, it probably seems less appealing to build in specific support for each DNS provider, but imho, that's the way to go if you want a maximally optimal UX for users of this feature.

That makes a lot of sense. I'll work on that. Thanks!

Comment (kevinjqiu) [reply stable/traefik/templates/deployment.yaml]: @krancour Want to run something by you before I implement it. The `dnsProvider` configuration will look like this:
```yaml
acme:
  enabled: true
  ...
  dnsProvider:
    route53:
      AWS_REGION: us-east-1
      AWS_ACCESS_KEY_ID: ...
      AWS_SECRET_ACCESS_KEY: ...
```

The `AWS_*` keys will be created as a kubernetes secret and injected into the pod via the deployment using env vars.

Does that sound like a good user interface for this feature?

Comment (krancour) [reply stable/traefik/templates/deployment.yaml]: I think you need a field still that lets the user specify _which_ of the many possible DNS providers to use, but overall, yes. This is very much what I had in mind. Whether now or in the future, more sections can be added for each provider. Both the secret and the deployment will have big `if / else if / else if / ... / else if` blocks to include the right fields in the secret and then for the deployment to grab the right env vars from the right fields in the secret.

What I like about this approach, as opposed to the map of environment variables, is that its more constrained. A glance at `values.yaml` (or `helm inspect traefik values`) is all you need to comprehend what options are available to you and how to set them. It's one-stop shopping instead of sending users out-of-band to peruse the Traefik docs.

👍

File: stable/traefik/values.yaml
Comment (krancour) [inline stable/traefik/values.yaml]: I thought we had discussed having a section per provider, with keys for relevant settings so that we weren't sending users straight off to the Traefik docs to figure out what do do. And then we'd need a discriminator of some sort to say which section to use.

**Comment by kevinjqiu** [comment]
Yes I'm working on it. I wanted to try a few contracts and see what works
best. I'll ping you when I get it to a reviewable state.

**Comment by kevinjqiu** [comment]
@krancour Updated. Could you please have a look? Tested with route53 and it still works.

**Comment by kevinjqiu** [comment]
@krancour :point_up: please.

File: stable/traefik/templates/deployment.yaml
Comment (krancour) [inline stable/traefik/templates/deployment.yaml]: I _think_ Go templates let you use `_` here in place of the `$v` you're not using.

**Comment by krancour** [comment]
@kevinjqiu this is _really_ nice. Fantastic work.

I have not test-driven it yet. I'd like to, but I'll be honest that I won't get to it  for a few days.

Are any of the other frequent contributors to this chart interested in validating?

cc @c-knowles @kachkaev

**Comment by krancour** [comment]
@unguiculus can we lose the stale tag, please?

File: stable/traefik/templates/deployment.yaml
Comment (kevinjqiu) [reply stable/traefik/templates/deployment.yaml]: Doesn't seem like it:
```
Error: parse error in "traefik/templates/deployment.yaml": template: traefik/templates/deployment.yaml:56: function "_" not defined
```
but I can replace it with `$_`.

Will change.

Comment (kevinjqiu) [reply stable/traefik/templates/deployment.yaml]: Fixed in https://github.com/kubernetes/charts/pull/1299/commits/2054d09db6f79a85932b51a89928a01158d1035a

**Comment by kevinjqiu** [comment]
@krancour @c-knowles @kachkaev 
guys what's the next step here?

File: stable/traefik/templates/deployment.yaml
Comment (cknowles) [inline stable/traefik/templates/deployment.yaml]: @krancour is there an agreement on how many versions back of k8s you want to support? Otherwise we could replace this range with `envFrom` that was introduced in k8s 1.6.

File: stable/traefik/README.md
Comment (cknowles) [inline stable/traefik/README.md]: If only one provider can be configured at a time, do we need to support each of these with separate yaml keys or could it be flattened to something like `acme.dnsProvider.envConfig`? I don't have a specific concern, just think it might be more complex than required.

**Review by cknowles — COMMENTED** [review]
Overall it seems good. We are not using this setup so I wasn't able to fully verify it. I have a couple of questions/comments.

File: stable/traefik/README.md
Comment (kevinjqiu) [reply stable/traefik/README.md]: @c-knowles 

> If only one provider can be configured at a time, do we need to support each of these with separate yaml keys or could it be flattened to something like acme.dnsProvider.envConfig

That's similar to how I [originally implemented](https://github.com/kubernetes/charts/pull/1299/commits/70386896e22963e189da195191f05fb4f2772359), but it's difficult to tie down the schema if we allow a blank `envConfig`. As @krancour pointed out, we want it to be clear to the user what keys are expected here instead of having them to consult the documentation of lego.

Another added benefit here is you can have multiple cloud DNS provider configured here but you may want to switch between them. This implementation allows the user to only change the `dnsProvider.name` field without touching anything else.

Comment (cknowles) [reply stable/traefik/README.md]: Ok sure, makes sense. I had missed the original changes. I think the keys could still be documented the other way as well but if the intention is that there’d eventually be a schema for the config then it’ll be easier like this with separate keys.

File: stable/traefik/templates/deployment.yaml
Comment (krancour) [reply stable/traefik/templates/deployment.yaml]: > @krancour is there an agreement on how many versions back of k8s you want to support?

Hadn't thought about it. Rather than force ourselves to figure that out right now, I'd say leave it as is.

**Comment by grugnog** [comment]
Noting that with the LE issue (https://community.letsencrypt.org/t/2018-01-09-issue-with-tls-sni-01-and-shared-hosting-infrastructure/49996) it appears that the SNI challenge that Traefik appears to use by default (at least in a Docker/Kubernetes context) is broken and will stay that way potentially for a while. See https://github.com/containous/traefik/issues/2691 for more details.

DNS challenge is a known workaround, so it would be nice to expedite this if possible (I'll give it a test myself shortly).

**Comment by grugnog** [comment]
I have tested this and confirmed that overall it is working (see log below), at least with Cloudflare. I needed to rebase it and resolve several merge conflicts and also make the values a map per-provider (rather than a slice of maps, which was failing to template). I also added quoting around the environment variables in the secret.

I have pushed my branch up to https://github.com/kubernetes/charts/compare/master...grugnog:traefik-dns-challenge

@kevinjqiu if you like, feel free to push my branch onto yours so we can keep the thread in this PR. Otherwise I am happy to open a new PR with my branch.

```
legolog: 2018/01/11 21:07:58 [INFO][-.-.-] acme: Obtaining bundled SAN certificate
legolog: 2018/01/11 21:07:58 [INFO][-.-.-] AuthURL: https://acme-v01.api.letsencrypt.org/acme/authz/-
legolog: 2018/01/11 21:07:58 [INFO][-.-.-] acme: Trying to solve DNS-01
legolog: 2018/01/11 21:07:58 [INFO][-.-.- Checking DNS record propagation using [-.-.-.-:-]
legolog: 2018/01/11 21:08:02 [INFO][-.-.-] The server validated our request
legolog: 2018/01/11 21:08:03 [INFO][-.-.-] acme: Validations succeeded; requesting certificates
legolog: 2018/01/11 21:08:33 [INFO] acme: Requesting issuer cert from https://acme-v01.api.letsencrypt.org/acme/issuer-cert
legolog: 2018/01/11 21:08:33 [INFO][-.-.-] Server responded with a certificate.
```

**Comment by kevinjqiu** [comment]
​Thanks for testing and fixing issues. I'll incorporate your changes
tonight :)

**Comment by kevinjqiu** [comment]
ping @krancour https://github.com/kubernetes/charts/pull/1299#issuecomment-357049450

File: stable/traefik/values.yaml
Comment (Azuka) [inline stable/traefik/values.yaml]: This should be gcloud, per: https://github.com/xenolf/lego/blob/master/providers/dns/dns_providers.go#L56

I tested after I cloned your branch.

File: stable/traefik/values.yaml
Comment (Azuka) [inline stable/traefik/values.yaml]: Is there a way to have this mounted as a volume? It looks like Google will only read from a file: https://github.com/golang/oauth2/blob/master/google/default.go#L56-L69

I'm not sure about other providers.

I'm going to try using the `rfc2136` option and see if Google supports that.

**Review by Azuka — COMMENTED** [review]
Awesome, and thanks for taking this on, @kevinjqiu.

I'm setting up a GKE cluster for the first time and this looks like the solution to getting acme working.

I noticed a couple things about the google cloud provider if you want to take a look.

File: stable/traefik/values.yaml
Comment (kevinjqiu) [reply stable/traefik/values.yaml]: Thanks. Fixed in https://github.com/kubernetes/charts/pull/1299/commits/6d6e4d71ed0733958bf967bd7fa5a27f91e5b9af.

**Comment by isaldarriaga** [comment]
Now that [Let's encrypt recently disabled TLS-SNI challenges](https://community.letsencrypt.org/t/tls-sni-challenges-disabled-for-most-new-issuance/50316), the DNS challenge looks like the option to take.

Is there any tutorial on how to make this feature work with the helm chart?

GKE
Godaddy
Traefik alpine 1.5/docker

Since i cannot renew certs in my test lab, i will disable acme in the meantime.

Many thanks.

**Comment by kevinjqiu** [comment]
@isaldarriaga Take a look at the README of the branch: https://github.com/kevinjqiu/charts/blob/traefik-dns-challenge/stable/traefik/README.md#lets-encrypt-domain-verification-using-dns-challenge

**Comment by kevinjqiu** [comment]
@krancour @c-knowles @Azuka Now that it appears there have been a significant amount of interest on this feature, can you consider merging this PR?

**Comment by jlk** [comment]
FWIW I've been running the PR locally in 2 k8s clusters for a few days with route53 DNS verification, seems to work well.

**Comment by kevinjqiu** [comment]
@krancour @c-knowles @unguiculus  Now that it appears there have been a significant amount of interest on this feature, can you consider merging this PR?

**Comment by krancour** [comment]
I think this is just about there. There's one thing I think does need to change, however...

There is another PR open for adding support for another challenge type (HTTP-01). Whether that's merged or not, it's clear that there are a variety of existing ACME challenge types now, and likely to be more in the future. As we bake support for multiple challenge types into this chart, there needs to be an unambiguous discriminator used to identify which challenge type should be used.

Currently, in this PR, the additional configuration in the traefik.toml (the config map) is guarded with this:

```
{{- if .Values.acme.dnsProvider.name }}
```

That might not be sufficient. We might want to consider something more like this:

```
{{- if and (eq .Values.acme.challengeType "DNS") .Values.acme.dnsProvider.name }}
```

And in values.yaml, the default value of `acme.challengeType` should reflect the current default. This should be overridden to opt into the DNS-based challenge (or any other challenge that is added in the future.)

**Comment by kevinjqiu** [comment]
@krancour updated as suggested. Two minor changes from #3347:

- `challengeType` is a grouped under `acme.challengeType` since all ACME related configs are there
- `http-01`/`dns-01` instead of the all-caps `HTTP-01`/`DNS-01` since that's consistent with how letsencrypt documentation references them

**Comment by kevinjqiu** [comment]
@krancour Updated as suggested.  Please have a look :point_up:

**Comment by Azuka** [comment]
@kevinjqiu, thanks again for all the work you've put into this. I wanted to bring up the Google Cloud options. Looks like the only way to supply credentials to `lego` is by mounting the credentials as a volume. I'm using these changes to the deployment under `volumes`:

```yaml
     {{- if eq "gcloud" .Values.acme.dnsProvider.name }}
      - name: gcloud
        secret:
          secretName: {{ template "fullname" $ }}-dnsprovider-config
          items:
          - key: GCE_SERVICE_ACCOUNT_FILE_DATA
            path: dns.json
    {{- end }}
```

and under `volumeMounts`

```yaml
        {{- if eq "gcloud" .Values.acme.dnsProvider.name }}
        - mountPath: /var/run/secrets/dns
          name: gcloud
        {{- end }}
```

My gcloud config in my passed in config file looks like the below:

```yaml
    gcloud:
      GCE_PROJECT: "<projectName>"
      GCE_SERVICE_ACCOUNT_FILE: '/var/run/secrets/dns/dns.json' #This doesn't work
      GOOGLE_APPLICATION_CREDENTIALS: '/var/run/secrets/dns/dns.json' #This does
      GCE_SERVICE_ACCOUNT_FILE_DATA: '{"type":"service_account","project_id":"<projectName>","private_key_id":"<hash>","private_key":"-----BEGIN PRIVATE KEY-----...","client_email":"...@<projectName>.iam.gserviceaccount.com","client_id":"...","auth_uri":"https://accounts.google.com/o/oauth2/auth","token_uri":"https://accounts.google.com/o/oauth2/token","auth_provider_x509_cert_url":"https://www.googleapis.com/oauth2/v1/certs","client_x509_cert_url":"https://www.googleapis.com/robot/v1/metadata/x509/kube-dns-admin%40<projectName>"}'
```

This feels a little clunky to me but works. Is this something we could incorporate into your PR, or would you prefer I submit a separate pull request after this is merged?

File: stable/traefik/values.yaml
Comment (krancour) [inline stable/traefik/values.yaml]: I don't quite understand how this section is being used.

**Review by krancour — CHANGES_REQUESTED** [review]

File: stable/traefik/values.yaml
Comment (krancour) [inline stable/traefik/values.yaml]: Isn't the default currently in use the TLS-SNI-01 challenge? The PR that adds support for the HTTP-01 challenge has not been merged yet.

File: stable/traefik/values.yaml
Comment (kevinjqiu) [reply stable/traefik/values.yaml]: This is already on master as of 05bd3660

File: stable/traefik/templates/configmap.yaml
Comment (krancour) [inline stable/traefik/templates/configmap.yaml]: It seems that `dnsProvider` is deprecated, per Traefik's docs:

```
# Use a DNS-01 acme challenge rather than TLS-SNI-01 challenge
#
# Optional (Deprecated, replaced by [acme.dnsChallenge])
#
# dnsProvider = "digitalocean"
```

```
# Use a DNS-01 acme challenge rather than TLS-SNI-01 challenge
#
# Optional
#
# [acme.dnsChallenge]

  # Provider used.
  #
  # Required
  #
  # provider = "digitalocean"

  # By default, the provider will verify the TXT DNS challenge record before letting ACME verify.
  # If delayBeforeCheck is greater than zero, avoid this & instead just wait so many seconds.
  # Useful if internal networks block external DNS queries.
  #
  # Optional
  # Default: 0
  #
  # delayBeforeCheck = 0
```

File: stable/traefik/values.yaml
Comment (krancour) [reply stable/traefik/values.yaml]: Ok... I thought it looked familiar. Can I suggest rebasing on master then? I see there are complementary changes to this that are absent from the deployment in this PR-- the end result being that anything in this section won't ever be included.

**Comment by kevinjqiu** [comment]
@Azuka Sorry for the late reply.

> Is this something we could incorporate into your PR, or would you prefer I submit a separate pull request after this is merged?

I think it's better if that could be a separate PR if it's ok with you. This one is already huge and has been open for quite a while.

**Comment by krancour** [comment]
@kevinjqiu, looking forward to getting this in. Are you able to address my few outstanding comments? Or need a hand?

**Comment by kevinjqiu** [comment]
@krancour I will have a go at it tonight. Hopefully will finish it by the end of the weekend.

File: stable/traefik/values.yaml
Comment (krancour) [reply stable/traefik/values.yaml]: I realize now that TLS-SNI-01 has been deprecated by Let's Encrypt. I understand why (vulnerabilities), but this is unfortunate. It nearly forces a breaking change on this chart.

For now, I'm going to suggest that we still treat TLS-SNI-01 as the default. (Traefik still supports it after all.) If it doesn't work, that's between the user and Let's Encrypt and isn't this chart's fault. (fwiw, if I understand correctly, it _will_ work for renewing existing certificates that were initially issued using that challenge.) i.e. We should, for now, treat this as "not our problem."

I think we can very soon consider a major revision of this chart and we can straighten this out better then.

**Comment by krancour** [comment]
@kevinjqiu I had some time on my hands today, so here's a pull request that might help you get this over the finish line:

https://github.com/kevinjqiu/charts/pull/1

**Comment by kevinjqiu** [comment]
@krancour Merged your change and sync'ed with master. Thanks!

**Comment by krancour** [comment]
@kevinjqiu looks like there is one remaining merge conflict. But this LGTM. If you can resolve that conflict, this should be merged.

**Comment by kevinjqiu** [comment]
@krancour Done.

**Comment by krancour** [comment]
Thanks @kevinjqiu.

**Comment by krancour** [comment]
LGTM.

@unguiculus any chance we can please get this merged today?

**Comment by kevinjqiu** [comment]
@unguiculus @krancour 

image

What do I have to do to make that build run? Or is that something in your control?

**Comment by krancour** [comment]
@kevinjqiu thanks for all the hard work on this! This is a great feature!

**Comment by kevinjqiu** [comment]
@krancour Thank you for the feedback!

**Pull request merged**
